Correlation-Heatmap Figures Creator : documentation
===================================================

Description
-----------

Those scripts aims to create correlation and Heatmaps graphics of exons regulated by several splicing factors.
Every data used to produce those graphics are taken from ``sed database``. Those script were used to produce the heatmaps and the correlation graphics in the article written by *Lemaire et al* entitled:

:align: center

  **Splicing factors and chromatin organization enhance exon recognition by alleviating constraints generated by gene nucleotide composition bias**


Heatmaps figures
~~~~~~~~~~~~~~~~

``heatmap_maker.py`` script aims, to cluster or simply sort exons activated by every **splicing factor** (defined in sed database) by their relative medians (or mean) for a features of interest.

The following formula was used to compute the relative median of a feature :math:`D` in a set of :math:`n` exons :math:`S` compared to a set of :math:`m` control exons :math:`C`.

.. math::

  RFreq(D) = \frac{median(D_{S}) - median(D_{C})}{median(D_{C})} \times 100

Where, :math:`D_S` and :math:`D_C` are the vectors of :math:`D` values for the sets :math:`S` and :math:`C`.

It is possible to create heamaps using the above formula but not with a :math:`median` function but a :math:`mean` function.

The control sets of exons are taken from sed and corresponds by default to every human consitutive coding exons (except those activated by the splicing factors defined in sed database).



Correlation figures
~~~~~~~~~~~~~~~~~~~

The goal of the ``correlation_maker.py`` script is to create a correlation figure between the relative median values of two features in every group exons activated or repressed by a splicing factor.

.. note::

  This script can also create that kind of figure at the **exon level** : it means that for a given group of exons, every exons in this group is repressented in this figure as a dot (whose coordinates corresponds to the values of the two represented features for the exon).
  The default group of exons correponds to every exons activated by at least one splicing factor



.. note::

  The features mentioned for **Correlation** and **Heatmaps** figures corresponds to the names of columns of the sed table in the sed database. To see the name of those features, you can read the documentation in the `Database_creator/` folder. The structure of sed database is displayed there.




Prerequisites
-------------

This program was written in **Ubuntu 16.04** with the version **3.5 of python**. It should work on other unix system with a bash shell.

Steps to launch those scripts:

  * You may have to install python 3, if you do click `here <https://www.python.org/downloads/release/python-356/>`_
  * Install the required dependencies by running ``sudo pip3 install -r requierements.txt``. The file ``requierements.txt`` is the folder ``Figure_ESA`` and contains the name and the version of every required module.
  * You must copy (or create a shortcut to) the sed database into the data folder of ``Figure_ESA``.


Then to know how to launch the programs creating the figures you can type:

.. code:: bash

  python3 src/heatmap_maker.py --help
  python3 src/correlation_maker.py --help



Executed code to produce the figures :

.. code:: bash

  # for heatmaps
  python3 src/heatmap_maker.py --name Fig_1A_iupac_exon --columns iupac_exon --union union --sf_type None --write_order Y --contrast 20 --nt S,W
  python3 src/heatmap_maker.py --name Fig_1C_min_flanking_intron_size --columns min_flanking_intron_size --union union --sf_type None --order /media/nicolas/DD_1/Splicing_Lore_project/Figure_ESA/result/new_heatmap_union/Fig_1A_iupac_exon_sorted.txt
  python3 src/heatmap_maker.py --name Fig_2B_left_panel_iupac_downstream_intron_25_nt_GC_exons --columns iupac_downstream_intron_adjacent1 --union union --sf_type GC_rich --order C --contrast 20 --operation mean
  python3 src/heatmap_maker.py --name Fig_2B_right_panel_iupac_downstream_intron_25_nt__AT_exons --columns iupac_downstream_intron_adjacent1 --union union --sf_type AT_rich --order C --contrast 20 --operation mean
  python3 src/heatmap_maker.py --name Fig_2C_left_panel_iupac_upstream_intron_25_nt_GC_exons --columns iupac_upstream_intron_adjacent1 --union union --sf_type GC_rich --order C --contrast 20 --operation mean
  python3 src/heatmap_maker.py --name Fig_2C_right_panel_iupac_upstream_intron_25_nt__AT_exons --columns iupac_upstream_intron_adjacent1 --union union --sf_type AT_rich --order C --contrast 20 --operation mean
  python3 src/heatmap_maker.py --name Fig_S3C_downstream_intron_iupac --columns iupac_downstream_intron --union union --sf_type None --order /media/nicolas/DD_1/Splicing_Lore_project/Figure_ESA/result/new_heatmap_union/Fig_1A_iupac_exon_sorted.txt --contrast 20 --nt S,W
  python3 src/heatmap_maker.py --name Fig_S3C_upstream_intron_iupac --columns iupac_upstream_intron --union union --sf_type None --order /media/nicolas/DD_1/Splicing_Lore_project/Figure_ESA/result/new_heatmap_union/Fig_1A_iupac_exon_sorted.txt --contrast 20 --nt S,W

  # for correlation figures
  python3 src/correlation_maker.py --xaxis min_flanking_intron_size --yaxis  iupac_exon --level exon --name Fig1D_correlation_S_exon_VS_min_flanking_intron_size --nt_list S --representation absolute
  python3 src/correlation_maker.py --xaxis iupac_exon --yaxis  iupac_upstream_intron --level exon --name Fig_1B_upper_correlation_S_exon_VS_S_upstream_intron --nt_list S --representation absolute
  python3 src/correlation_maker.py --xaxis iupac_exon --yaxis  iupac_downstream_intron --level exon --name Fig_1B_lower_correlation_S_exon_VS_S_downstream_intron --nt_list S --representation absolute
  python3 src/correlation_maker.py --xaxis iupac_exon --yaxis  iupac_gene --level exon --name Fig_4A_iupac_exon_vs_iupac_gene --nt_list S --exon_class GC-AT --representation absolute
  python3 src/correlation_maker.py --xaxis iupac_gene --yaxis  gene_size --level exon --name Fig_S4B_gene_size_vs_iupac_gene --nt_list S --exon_class GC-AT
